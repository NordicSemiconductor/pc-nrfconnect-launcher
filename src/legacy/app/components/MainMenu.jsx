/* Copyright (c) 2015 - 2017, Nordic Semiconductor ASA
 *
 * All rights reserved.
 *
 * Use in source and binary forms, redistribution in binary form only, with
 * or without modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions in binary form, except as embedded into a Nordic
 *    Semiconductor ASA integrated circuit in a product or a software update for
 *    such product, must reproduce the above copyright notice, this list of
 *    conditions and the following disclaimer in the documentation and/or other
 *    materials provided with the distribution.
 *
 * 2. Neither the name of Nordic Semiconductor ASA nor the names of its
 *    contributors may be used to endorse or promote products derived from this
 *    software without specific prior written permission.
 *
 * 3. This software, with or without modification, must only be used with a Nordic
 *    Semiconductor ASA integrated circuit.
 *
 * 4. Any software provided in binary form under this license must not be reverse
 *    engineered, decompiled, modified and/or disassembled.
 *
 * THIS SOFTWARE IS PROVIDED BY NORDIC SEMICONDUCTOR ASA "AS IS" AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY, NONINFRINGEMENT, AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL NORDIC SEMICONDUCTOR ASA OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
 * TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import React from 'react';
import Dropdown from 'react-bootstrap/Dropdown';
import Immutable from 'immutable';
import { instanceOf, oneOfType, string } from 'prop-types';

import MenuItem from '../../components/HotkeyedMenuItem';

const renderItems = menuItems =>
    menuItems.map(({ id, onClick, isDivider, hotkey, text }) => (
        <MenuItem
            key={id}
            onClick={onClick}
            divider={isDivider}
            hotkey={hotkey ? hotkey.toLowerCase() : ''}
            title={hotkey ? `${text} (${hotkey})` : text}
        >
            {text}
        </MenuItem>
    ));

const MainMenu = ({
    menuItems,
    iconName,
    cssClass,
    dropdownCssClass,
    dropdownMenuCssClass,
    ...rest
}) => (
    <div className={cssClass}>
        <Dropdown id="main-menu" {...rest}>
            <Dropdown.Toggle className={dropdownCssClass}>
                <span className={iconName} />
            </Dropdown.Toggle>
            <Dropdown.Menu id="main-menu-list" className={dropdownMenuCssClass}>
                {renderItems(menuItems)}
            </Dropdown.Menu>
        </Dropdown>
    </div>
);

MainMenu.propTypes = {
    menuItems: oneOfType([instanceOf(Array), instanceOf(Immutable.Iterable)])
        .isRequired,
    iconName: string,
    cssClass: string,
    dropdownCssClass: string,
    dropdownMenuCssClass: string,
};

MainMenu.defaultProps = {
    iconName: 'mdi mdi-menu',
    cssClass: 'core-nav-section core-padded-row',
    dropdownCssClass: 'core-main-menu core-btn btn-primary',
    dropdownMenuCssClass: 'core-dropdown-menu',
};

export default MainMenu;
