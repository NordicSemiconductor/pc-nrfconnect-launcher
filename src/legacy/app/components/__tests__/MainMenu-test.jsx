/* Copyright (c) 2015 - 2017, Nordic Semiconductor ASA
 *
 * All rights reserved.
 *
 * Use in source and binary forms, redistribution in binary form only, with
 * or without modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions in binary form, except as embedded into a Nordic
 *    Semiconductor ASA integrated circuit in a product or a software update for
 *    such product, must reproduce the above copyright notice, this list of
 *    conditions and the following disclaimer in the documentation and/or other
 *    materials provided with the distribution.
 *
 * 2. Neither the name of Nordic Semiconductor ASA nor the names of its
 *    contributors may be used to endorse or promote products derived from this
 *    software without specific prior written permission.
 *
 * 3. This software, with or without modification, must only be used with a Nordic
 *    Semiconductor ASA integrated circuit.
 *
 * 4. Any software provided in binary form under this license must not be reverse
 *    engineered, decompiled, modified and/or disassembled.
 *
 * THIS SOFTWARE IS PROVIDED BY NORDIC SEMICONDUCTOR ASA "AS IS" AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY, NONINFRINGEMENT, AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL NORDIC SEMICONDUCTOR ASA OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
 * TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/* eslint-disable import/first */

// Do not decorate components
jest.mock('../../../decoration', () => ({
    decorate: component => component,
}));

import React from 'react';
import { mount } from 'enzyme';
import Immutable from 'immutable';

import MainMenu from '../MainMenu';

const menuItems = Immutable.List([
    {
        id: 1,
        text: 'First item',
    },
    {
        id: 2,
        isDivider: true,
    },
    {
        id: 3,
        text: 'Last item',
    },
]);

describe('MainMenu', () => {
    it('should render menu with no items', () => {
        expect(
            mount(<MainMenu menuItems={[]} defaultShow />)
        ).toMatchSnapshot();
    });

    it('should render menu with two items separated by divider', () => {
        expect(
            mount(<MainMenu menuItems={menuItems} defaultShow />)
        ).toMatchSnapshot();
    });

    it('should invoke onClick when item has been clicked', () => {
        const onClick = jest.fn();
        const wrapper = mount(
            <MainMenu
                menuItems={[
                    {
                        id: 1,
                        text: 'Foo',
                        onClick,
                    },
                ]}
                defaultShow
            />
        );
        wrapper.find('a[title="Foo"]').first().simulate('click');

        expect(onClick).toHaveBeenCalled();
    });
});
