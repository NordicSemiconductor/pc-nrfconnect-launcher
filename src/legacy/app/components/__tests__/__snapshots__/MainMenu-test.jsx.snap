// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainMenu should render menu with no items 1`] = `
<MainMenu
  cssClass="core-nav-section core-padded-row"
  defaultShow={true}
  dropdownCssClass="core-main-menu core-btn btn-primary"
  dropdownMenuCssClass="core-dropdown-menu"
  iconName="mdi mdi-menu"
  menuItems={Array []}
>
  <div
    className="core-nav-section core-padded-row"
  >
    <Dropdown
      defaultShow={true}
      id="main-menu"
      navbar={false}
    >
      <ReactOverlaysDropdown
        itemSelector=".dropdown-item:not(.disabled):not(:disabled)"
        onToggle={[Function]}
        show={true}
      >
        <div
          className="show dropdown"
          id="main-menu"
        >
          <DropdownToggle
            className="core-main-menu core-btn btn-primary"
          >
            <Button
              active={false}
              aria-expanded={true}
              aria-haspopup={true}
              className="core-main-menu core-btn btn-primary dropdown-toggle"
              disabled={false}
              onClick={[Function]}
              variant="primary"
            >
              <button
                aria-expanded={true}
                aria-haspopup={true}
                className="core-main-menu core-btn btn-primary dropdown-toggle btn btn-primary"
                disabled={false}
                onClick={[Function]}
                type="button"
              >
                <span
                  className="mdi mdi-menu"
                />
              </button>
            </Button>
          </DropdownToggle>
          <DropdownMenu
            align="left"
            alignRight={false}
            className="core-dropdown-menu"
            flip={true}
            id="main-menu-list"
          >
            <div
              aria-labelledby=""
              className="core-dropdown-menu dropdown-menu show"
              id="main-menu-list"
              style={
                Object {
                  "left": "0",
                  "opacity": "0",
                  "pointerEvents": "none",
                  "position": "absolute",
                  "top": "0",
                }
              }
              x-placement="bottom-start"
            />
          </DropdownMenu>
        </div>
      </ReactOverlaysDropdown>
    </Dropdown>
  </div>
</MainMenu>
`;

exports[`MainMenu should render menu with two items separated by divider 1`] = `
<MainMenu
  cssClass="core-nav-section core-padded-row"
  defaultShow={true}
  dropdownCssClass="core-main-menu core-btn btn-primary"
  dropdownMenuCssClass="core-dropdown-menu"
  iconName="mdi mdi-menu"
  menuItems={
    Immutable.List [
      Object {
        "id": 1,
        "text": "First item",
      },
      Object {
        "id": 2,
        "isDivider": true,
      },
      Object {
        "id": 3,
        "text": "Last item",
      },
    ]
  }
>
  <div
    className="core-nav-section core-padded-row"
  >
    <Dropdown
      defaultShow={true}
      id="main-menu"
      navbar={false}
    >
      <ReactOverlaysDropdown
        itemSelector=".dropdown-item:not(.disabled):not(:disabled)"
        onToggle={[Function]}
        show={true}
      >
        <div
          className="show dropdown"
          id="main-menu"
        >
          <DropdownToggle
            className="core-main-menu core-btn btn-primary"
          >
            <Button
              active={false}
              aria-expanded={true}
              aria-haspopup={true}
              className="core-main-menu core-btn btn-primary dropdown-toggle"
              disabled={false}
              onClick={[Function]}
              variant="primary"
            >
              <button
                aria-expanded={true}
                aria-haspopup={true}
                className="core-main-menu core-btn btn-primary dropdown-toggle btn btn-primary"
                disabled={false}
                onClick={[Function]}
                type="button"
              >
                <span
                  className="mdi mdi-menu"
                />
              </button>
            </Button>
          </DropdownToggle>
          <DropdownMenu
            align="left"
            alignRight={false}
            className="core-dropdown-menu"
            flip={true}
            id="main-menu-list"
          >
            <div
              aria-labelledby=""
              className="core-dropdown-menu dropdown-menu show"
              id="main-menu-list"
              style={
                Object {
                  "left": "0",
                  "opacity": "0",
                  "pointerEvents": "none",
                  "position": "absolute",
                  "top": "0",
                }
              }
              x-placement="bottom-start"
            >
              <HotkeyedMenuItem
                divider={false}
                hotkey=""
                key="1"
                onClick={[Function]}
                title="First item"
              >
                <DropdownItem
                  as={
                    Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "displayName": "SafeAnchor",
                      "render": [Function],
                    }
                  }
                  disabled={false}
                  hotkey=""
                  onClick={[Function]}
                  title="First item"
                >
                  <SafeAnchor
                    className="dropdown-item"
                    disabled={false}
                    hotkey=""
                    onClick={[Function]}
                    title="First item"
                  >
                    <a
                      className="dropdown-item"
                      hotkey=""
                      href="#"
                      onClick={[Function]}
                      onKeyDown={[Function]}
                      role="button"
                      title="First item"
                    >
                      First item
                    </a>
                  </SafeAnchor>
                </DropdownItem>
              </HotkeyedMenuItem>
              <HotkeyedMenuItem
                divider={true}
                hotkey=""
                key="2"
                onClick={[Function]}
              >
                <DropdownDivider
                  role="separator"
                >
                  <div
                    className="dropdown-divider"
                    role="separator"
                  />
                </DropdownDivider>
              </HotkeyedMenuItem>
              <HotkeyedMenuItem
                divider={false}
                hotkey=""
                key="3"
                onClick={[Function]}
                title="Last item"
              >
                <DropdownItem
                  as={
                    Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "displayName": "SafeAnchor",
                      "render": [Function],
                    }
                  }
                  disabled={false}
                  hotkey=""
                  onClick={[Function]}
                  title="Last item"
                >
                  <SafeAnchor
                    className="dropdown-item"
                    disabled={false}
                    hotkey=""
                    onClick={[Function]}
                    title="Last item"
                  >
                    <a
                      className="dropdown-item"
                      hotkey=""
                      href="#"
                      onClick={[Function]}
                      onKeyDown={[Function]}
                      role="button"
                      title="Last item"
                    >
                      Last item
                    </a>
                  </SafeAnchor>
                </DropdownItem>
              </HotkeyedMenuItem>
            </div>
          </DropdownMenu>
        </div>
      </ReactOverlaysDropdown>
    </Dropdown>
  </div>
</MainMenu>
`;
